container_linux.go

```diff
@@ -252,6 +260,124 @@ func (c *linuxContainer) NotifyOOM() (<-chan struct{}, error) {
+	if e := os.Getenv(evar); e != "" {
+		for _, f := range strings.Fields(e) {
+			*args = append(*args, f)
+		}
+	}
+	fmt.Printf(">>> criu %v\n", *args)
+}
+
+func (c *linuxContainer) Checkpoint() error {
+	c.m.Lock()
+	defer c.m.Unlock()
+	dir := filepath.Join(c.root, "checkpoint")
+	// Since a container can be C/R'ed multiple times,
+	// the checkpoint directory may already exist.
+	if err := os.Mkdir(dir, 0655); err != nil && !os.IsExist(err) {
+		return err
+	}
+	args := []string{
+		"dump", "-v4",
+		"-D", dir,
+		"-o", "dump.log",
+		"--root", c.config.Rootfs,
+		"--manage-cgroups", "--evasive-devices",
+		"-t", strconv.Itoa(c.initProcess.pid()),
+	}
+	for _, m := range c.config.Mounts {
+		if m.Device == "bind" {
+			args = append(args,
+				"--ext-mount-map", fmt.Sprintf("%s:%s", m.Destination, m.Destination))
+		}
+	}
+	addArgsFromEnv("CRIU_C", &args) // XXX debug
+	if err := exec.Command(c.criuPath, args...).Run(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *linuxContainer) Restore(process *Process) error {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	pidfile := filepath.Join(c.root, "restoredpid")
+	// Make sure pidfile doesn't already exist from a
+	// previous restore.  Otherwise, CRIU will fail.
+	if err := os.Remove(pidfile); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	// XXX We should do the restore in detached mode (-d).
+	//     To do this, we need an "init" process that executes
+	//     CRIU and waits for it, reaping its children, and
+	//     waiting for the container.
+	args := []string{
+		"restore", "-v4",
+		"-D", filepath.Join(c.root, "checkpoint"),
+		"-o", "restore.log",
+		"--root", c.config.Rootfs,
+		"--pidfile", pidfile,
+		"--manage-cgroups", "--evasive-devices",
+	}
+	for _, m := range c.config.Mounts {
+		if m.Device == "bind" {
+			args = append(args, "--ext-mount-map",
+				fmt.Sprintf("%s:%s", m.Destination, m.Source))
+		}
+	}
+	// Pipes that were previously set up for std{in,out,err}
+	// were removed after checkpoint.  Use the new ones.
+	for i := 0; i < 3; i++ {
+		if s := c.config.StdFds[i]; strings.Contains(s, "pipe:") {
+			args = append(args, "--inherit-fd", fmt.Sprintf("fd[%d]:%s", i, s))
+		}
+	}
+	addArgsFromEnv("CRIU_R", &args) // XXX debug
+
+	// XXX This doesn't really belong here as our caller should have
+	//     already set up root (including devices) and mounted it.
+	/*
+		// remount root for restore
+		if err := syscall.Mount(c.config.Rootfs, c.config.Rootfs, "bind", syscall.MS_BIND|syscall.MS_REC, ""); err != nil {
+			return err
+		}
+
+		defer syscall.Unmount(c.config.Rootfs, syscall.MNT_DETACH)
+	*/
+	cmd := exec.Command(c.criuPath, args...)
+	cmd.Stdin = process.Stdin
+	cmd.Stdout = process.Stdout
+	cmd.Stderr = process.Stderr
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	// cmd.Wait() waits cmd.goroutines which are used for proxying file descriptors.
+	// Here we want to wait only the CRIU process.
+	st, err := cmd.Process.Wait()
```

Comment:
I'm working on getting this branch up and running in Docker, but I'm a bit unsure about what to do here. Because we're not passing the -d flag to criu, this Wait will actually just wait forever -- that's fine in the sort of direct nsinit resume an interactive container case (which I had used to test this branch), but it doesn't work when dealing with the Docker daemon. Should this be a configurable flag we're passing in, or should the wait be handled at the nsinit level and not here?